Firmware/tmc2130.cpp

```diff
@@ -883,14 +885,115 @@ void tmc2130_get_wave(uint8_t axis, uint8_t* data)
+// - fac: power factor for linearity correction
+// - tcorr: pre-calculated torque correction factor
+// - carry: quantization error carry-forward (modified by reference)
+// - prev_theoretical_value: previous theoretical value for slope calculation (modified by reference)
+//
+// Returns: 8-bit amplitude value clamped to [SIN0, AMP] range
+//
+// References:
+// - Prusa Forum: "TMC2130 constant torque algorithm" discussion
+//   https://forum.prusa3d.com/forum/original-prusa-i3-mk3s-mk3-user-mods-octoprint-enclosures-nozzles/stepper-motor-upgrades-to-eliminate-vfa-s-vertical-fine-artifacts/paged/2/
+// - Analog Devices AN-026: "Stepper Motor Control Using TMC2130"
+//   https://www.analog.com/en/resources/app-notes/an-026.html
+uint8_t tmc2130_calc_constant_torque_value(uint8_t i, uint8_t va, float fac, float tcorr,
+                                          float& prev_theoretical_value) {
+	constexpr uint8_t SIN0 = 0;
+	constexpr uint8_t AMP = 248;  // Amplitude limit as per AN-026 recommendation
+	constexpr float TARGET_MAGNITUDE_SQUARED = (float)AMP * AMP + (float)SIN0 * SIN0;
+
+	// Theoretical constant torque value at microstep position i
+	float theoretical_value;
+
+	if (i < 128) {
+		// Phase 1 (positions 0-127): Power-corrected sine curve
+		// Calculate theoretical value using sine function with power factor
+		// correction and tcorr adjustment for midpoint matching
+		float sin_val = sin(M_PI * (float)i / 512.0f);
+		theoretical_value = (AMP - SIN0) * pow(sin_val, fac) * tcorr + SIN0;
+	} else {
+		// Phase 2 (positions 128-255): Constant torque constraint solving
+		// For constant torque: |A(i)|² + |B(i)|² = TARGET_MAGNITUDE_SQUARED
+		// Since B(i) = A(255-i), solve: |A(i)|² + |A(255-i)|² = TARGET_MAGNITUDE_SQUARED
+		// Therefore: A(i) = sqrt(TARGET_MAGNITUDE_SQUARED - A(255-i)²)
+
+		// Calculate mirror position value from Phase 1 curve
+		uint8_t mirror_i = 255 - i;
+		float sin_val = sin(M_PI * (float)mirror_i / 512.0f);
+		float mirror_theoretical = (AMP - SIN0) * pow(sin_val, fac) * tcorr + SIN0;
+
+		// Apply constant torque constraint
+		theoretical_value = sqrt(TARGET_MAGNITUDE_SQUARED - mirror_theoretical * mirror_theoretical);
+	}
+
+	// Step 1: Initial quantization using simple rounding
+	uint8_t candidate_value = (uint8_t)(theoretical_value + 0.5);
+
+	// Step 2: Slope-based delta limiting for TMC2130 compression
+	// Calculate slope between current and previous theoretical values
+	float slope = theoretical_value - prev_theoretical_value;
+
+	// Determine allowed delta range
+	// This ensures delta ranges match slope ranges for optimal compression:
+	// slope ∈ [0,1) → deltas ∈ [0,1], slope ∈ [1,2) → deltas ∈ [1,2], etc.
+	int8_t min_delta = (int8_t)floor(slope);
+
+	// Clamp to TMC2130 hardware delta limits [-1, 3]
+	// Max delta is 2 because we need range [min_delta, min_delta+1]
+	if (min_delta < -1) {
+		min_delta = -1;
+	} else if (min_delta > 2) {
+		min_delta = 2;
```

Comment:
Sometimes we can save some memory by using the constrain macro. 

min_delta = constrain(min_delta, -1, 2);

It wouldnt save much but perhaps some.